diff --git a/iCalcreator.class.php b/iCalcreator.class.php
index 291c398..92379ea 100755
--- a/iCalcreator.class.php
+++ b/iCalcreator.class.php
@@ -73,7 +73,7 @@ class vcalendar {
  * @param array $config
  * @return void
  */
-  function vcalendar ( $config = array()) {
+  function __construct( $config = array()) {
     $this->_makeVersion();
     $this->calscale   = null;
     $this->method     = null;
@@ -1969,7 +1969,7 @@ class calendarComponent {
  * @author Kjell-Inge Gustafsson, kigkonsult <ical@kigkonsult.se>
  * @since 2.9.6 - 2011-05-17
  */
-  function calendarComponent() {
+  function __construct() {
     $this->objName         = ( isset( $this->timezonetype )) ?
                           strtolower( $this->timezonetype )  :  get_class ( $this );
     $this->uid             = array();
@@ -6230,8 +6230,8 @@ class vevent extends calendarComponent {
  * @param  array $config
  * @return void
  */
-  function vevent( $config = array()) {
-    $this->calendarComponent();
+  function __construct( $config = array()) {
+    parent::__construct();
 
     $this->attach          = '';
     $this->attendee        = '';
@@ -6377,8 +6377,8 @@ class vtodo extends calendarComponent {
  * @param array $config
  * @return void
  */
-  function vtodo( $config = array()) {
-    $this->calendarComponent();
+  function __construct( $config = array()) {
+    parent::__construct();
 
     $this->attach          = '';
     $this->attendee        = '';
@@ -6516,8 +6516,8 @@ class vjournal extends calendarComponent {
  * @param array $config
  * @return void
  */
-  function vjournal( $config = array()) {
-    $this->calendarComponent();
+  function __construct( $config = array()) {
+    parent::__construct();
 
     $this->attach          = '';
     $this->attendee        = '';
@@ -6626,8 +6626,8 @@ class vfreebusy extends calendarComponent {
  * @param array $config
  * @return void
  */
-  function vfreebusy( $config = array()) {
-    $this->calendarComponent();
+  function __construct( $config = array()) {
+    parent::__construct();
 
     $this->attendee        = '';
     $this->comment         = '';
@@ -6708,8 +6708,8 @@ class valarm extends calendarComponent {
  * @param array $config
  * @return void
  */
-  function valarm( $config = array()) {
-    $this->calendarComponent();
+  function __construct( $config = array()) {
+    parent::__construct();
 
     $this->action          = '';
     $this->attach          = '';
@@ -6791,7 +6791,7 @@ class vtimezone extends calendarComponent {
  * @param array $config
  * @return void
  */
-  function vtimezone( $timezonetype=FALSE, $config = array()) {
+  function __construct( $timezonetype=FALSE, $config = array()) {
     if( is_array( $timezonetype )) {
       $config       = $timezonetype;
       $timezonetype = FALSE;
@@ -6800,7 +6800,7 @@ class vtimezone extends calendarComponent {
       $this->timezonetype = 'VTIMEZONE';
     else
       $this->timezonetype = strtoupper( $timezonetype );
-    $this->calendarComponent();
+    parent::__construct();
 
     $this->comment         = '';
     $this->dtstart         = '';
